### PR TITLE
[Full DTensor] DTensor-fy the inputs and support simple fsdp with dtensor activation and loss

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #2049

**Full Dtensor PRs will be landed to full_dtensor branch**

**Summary**
This PR introduces an initial prototype and skeleton for fully DTensor-based training. The current codebase builds upon SimpleFSDP functions. However, we reimplement how we apply FSDP/DDP in this PR. The new approach mimics the semantics that each module provides the knowledge of how to parallelize the states (parameters and buffers). 

**TODOs (order doesn't matter)**
[x] DTensor-fy the inputs and support simple fsdp with dtensor activation and loss
[ ] register_parametrization that directly handle the modules parameters loop and shared parameters
[ ] Enable TP for llama3
[ ] Enable CP for llama3
[ ] local_map inner_attention (FlexAttention + DTensor is not completely ready)
[ ] Using seed checkpoint gives us a different loss curve. Need to understand if this is a checkpointing bug
[ ] Make fully_shard work with full dtensor (need some PRs in PyTorch)
[ ] Enable PP for llama3
[ ] Enable AC wrapper for llama3 
[ ] Integrate with the new configuration system
[ ] Investigate how to remove build_parallelize_inputs_fn and make it be a part of parallelizing a module.
[ ] Enable FlexAttention + DTensor once the PR in PyTorch is ready
[ ] DSV3 (more items to extend later).

**Accuracy verification:**
SimpleFSDP v.s. FSDP2

```
python3 scripts/loss_compare.py . . \
--baseline-options='--activation_checkpoint.mode="none" --parallelism.data_parallel_replicate_degree=2' \
--test-options='--model.name full_dtensor.llama3 --activation_checkpoint.mode="none" --parallelism.data_parallel_replicate_degree=2' \
--test-train-file=torchtitan.experiments.full_dtensor.train  \
--steps=10 --assert-equal --no-seed-checkpoint
...
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

### CODE REVIEW

**Comment by wconstab** [comment]
is this at a good point for review?  / do you have thoughts on what the next steps and long term design are?

**Comment by tianyu-l** [comment]
From last meeting, we sort of agreed on merging multiple experiments folder into one, including compiler_tookit, simplefsdp, and maybe autoparallel. I would suggest we take this opportunity to consolidate all the effort.

**Comment by tianyu-l** [comment]
Also I'm planning to redo config, which may need to go hand-in-hand with how parallelisms are applied.

**Comment by fegin** [comment]
> is this at a good point for review? / do you have thoughts on what the next steps and long term design are?

@wconstab Not yet, I'll publish once some big pictures are finalized. For example, the question below, are we okay to have a separate experimental folder during the development. 

> From last meeting, we sort of agreed on merging multiple experiments folder into one, including compiler_tookit, simplefsdp, and maybe autoparallel. I would suggest we take this opportunity to consolidate all the effort.

@tianyu-l 
Do we still want TorchTitan's simple_fsdp to support non-full-dtensor case? If not, I think it does make sense to combine all. 

However, as you mentioned,
> Also I'm planning to redo config, which may need to go hand-in-hand with how parallelisms are applied.

We will probably break couple xD from time to time when doing the refactor of configuration and full_dtensor. So it may be good to have a temporary separate folder during this development. Once we enable all parallelisms, we can merge it back to whatever centralized experimental folder we have.

**Comment by tianyu-l** [comment]
> We will probably break couple xD from time to time when doing the refactor of configuration and full_dtensor. So it may be good to have a temporary separate folder during this development. Once we enable all parallelisms, we can merge it back to whatever centralized experimental folder we have.

Agreed on the refactor part. Would development in a branch an acceptable strategy?

File: torchtitan/experiments/full_dtensor/train.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from typing import Any
+
+import torch
+
+from torchtitan.train import main, Trainer
+
+from .llama3.parallelize import build_parallelize_inputs_fn, parallelize_buffers
+
+
+class FullDTensorTrainer(Trainer):
```
Comment (zpcore) [inline torchtitan/experiments/full_dtensor/train.py:17]: How does this Trainer get called?

Comment (zpcore) [reply torchtitan/experiments/full_dtensor/train.py:17]: Oh, I see, it's trigger automatically by `torchtitan.experiments.full_dtensor.train`.

**Comment by fegin** [comment]
Per the discussion with @tianyu-l, I created a full_dtensor branch. I'll land the PR (and the following PRs) to full_dtensor branch. cc., @zpcore

**Review by tianyu-l — APPROVED** [review]
